lib/Doctrine/DBAL/Driver/PDOException.php

```diff
@@ -9,10 +9,7 @@
  */
 class PDOException extends AbstractDriverException
 {
-    /**
-     * @param \PDOException $exception The PDO exception to wrap.
-     */
-    public function __construct(\PDOException $exception)
+    public static function fromNativePDOException(\PDOException $exception) : PDOException
```

Comment:
This is something I was also trying to do as part of my other patch which is no longer relevant. I'll review this later once I check my corresponding notes.

Reply:
There is so much room for improvement in so many places, that it's very difficult to organize PRs that don't have dependencies, (when not circular dependencies!) on other PRs, sometimes conflicting ones from other people. Hence my previous comments.

Reply:
I agree. I'll make sure we're not losing control over the scope of the patch.

Reply:
> I agree. I'll make sure we're not losing control over the scope of the patch.

Still considering the above, I don't think that this is an acceptable approach. This is yet another case that shows that the current driver exception hierarchy is not maintainable.

I cannot let introduce this method because having it will make master non-releasable until it's removed. Or otherwise, we'd have to maintain this method until the next major release.

With that said, given all other revealed driver exception issues, why don't we focus on that and get back to this project later? Otherwise, it turns into a rabbit hole and requires significant mental effort to keep all considerations in mind.

Reply:
> This is something I was also trying to do as part of my other patch which is no longer relevant.

I was unable to find it but it was similar to what you're trying to do.

Reply:
Oh, I wasn't imagining releasing 3.0 *before* reworking the driver exceptions. But we can work on them before if you wish (that was actually the plan in #3368).

If you're OK with this, I'll start with making `DriverException` a single concrete class, and drop the current subclasses / implementations. This will basically be a rebased version of #3367, which is totally out of date now.

We'll then be able to proceed to subclassing `DriverException` for specific *exception reasons*, not *drivers*, later, if required.

Reply:
> `DriverException` is already a class and not an interface anymore.

I must have missed it. When did it happen? Again, we have two symbols with the same name `DriverException`:
```sh
$ git grep -P "(class|interface) DriverException"
lib/Doctrine/DBAL/Driver/DriverException.php:interface DriverException extends Throwable
lib/Doctrine/DBAL/Exception/DriverException.php:class DriverException extends DBALException implements DriverExceptionInterface
```

> I will just try to remove driver-specific exceptions and move factory methods to driver connections, which was the original aim. Is that potentially OK with you?

That sounds about right. Ideally, we don't want to have any factory methods but named exceptions instead. E.g. instead of `OCI8Exception::fromErrorInfo()` we may want to have `OCI8Error::new()`. Let's start with a prototype and discuss it on the code level. As I said earlier, I don't have the full picture in my head.

Reply:
How is `OCIError` different from `OCI8Exception`? My idea is to *remove* `OCI8Exception`, not rename it! Thus the need to move the factory method elsewhere; and I think it belongs in the connection or the statement that throws it.

Reply:
The idea is that `OCI8Error` would represent only an error that happened because of `oci_error()`. Otherwise, how would you translate it to an exception? No other errors that have happened in the DBAL's `OCI8` driver will have to be represented with this class. Probably, it could use a better name if it turns out to be confusing.

In any event, I'm not suggesting anything specific, go ahead and draft your vision, and we can discuss it.

Reply:
What I mean is that *when* we need factory methods for driver-specific exceptions, we should avoid creating subclasses specific to that driver just for the purpose of putting factory methods there. Having an `OC8IException`|`Error` class would imply that you might need to catch it at some point, when in reality you're only interested in catching the base `DriverException`. Therefore I'd move the factory methods elsewhere.

I'll draft a PR soon.

Reply:
> Having an `OC8IException`|`Error` class would imply that you might need to catch it at some point.

Why would it imply that? It's just an implementation of the interface that the throwing method would declare in its `@throws` annotation, and this is what you need to be catching.

Reply:
Because drivers implement interfaces, such as `Driver\Connection`, that explicitly declare which exceptions may be thrown.

`Connection::commit()`, for example, declares throwing a `DriverException`. As a consumer of this interface, you should only be interested in catching this exception, you're not supposed to dig into the internals to find out that there is a (non portable) `OCI8Exception` lying somewhere. Sure, it doesn't *hurt* to have this class internally, as it does extend `DriverException`, but this leaves the door open to someone writing a `catch (OCI8Exception)` in their code, which would be a big code smell, as this would defeat portability, which is the whole point of the DBAL.

Therefore **I'd better not have any driver-specific exception**, and only subclass `DriverException` if there is a legitimate, *cross-driver* use case for it.

If you look at current driver-specific exceptions, you'll notice that **they're only there to provide driver-specific factory methods**. This is not a big enough justification to subclass `DriverException`; the factory methods may live elsewhere.